### PR TITLE
KHI_growthRate Testsuite

### PR DESCRIPTION
**Pull request by MSVoss**
Integrate the test suite that compares the growth rates for the Kelvin Helmholtz instability setup

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
The python scripts are not pep8 formated: https://gitlab.com/hzdr/crp/picongpu/-/jobs/2453910558
All python scripts need to pass pep8 code style checks and flake8 checks.

File: test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+if [ $PIC_RET != 0 ]; then
+    echo "running PIConGPU failed with: " $PIC_RET
+    exit $PIC_RET
+fi
+
+# running the testsuite
+MAINTEST="$TBG_dstPath/input/lib/python/test/KHI_growthRate"
+cd $TBG_dstPath/simOutput/
+python $MAINTEST/MainTest.py $MAINTEST $TBG_dstPath/input/include/picongpu/param/ $TBG_dstPath/simOutput/
```
Comment (psychocoderHPC) [inline test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl]: We should not introduce a `tpl` file in examples because these file will be soon outdated.
Could you please provide a `test.sh` bash script (do not forget to set execution rights) in the root of the input set `KHI_growthRate` which returns the error code of the python script e.g.
```
#!/bin/bash
MAINTEST=input/lib/python/test/KHI_growthRate"
cd simOutput/
python $MAINTEST/MainTest.py $MAINTEST ../input/include/picongpu/param/ .
exit $?
```

I assume that the python script returns zero in case the test passed and non zero else.

I will add in a separate PR the short code snipped that copies `test.sh` when calling `tbg`

File: lib/python/test/KHI_growthRate/MainTest.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	plotter = ViewerKHI(paramDir, simDir)
+	plotter.plotGrowthRate(field, True, True, True)
+
+except:	
+	error0 = str(sys.exc_info()[0])
+	error1 = str(sys.exc_info()[1])
+	error2 = str(sys.exc_info()[2])	
+
+	#print error0 + error1 + error2
+	fobj_out = open(sys.argv[3]+ "/fehler.log","w")
```
Comment (psychocoderHPC) [inline lib/python/test/KHI_growthRate/MainTest.py]: please use english terms `e.g. khi_error.log`

File: lib/python/test/KHI_growthRate/MainTest.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import sys
```
Comment (psychocoderHPC) [inline lib/python/test/KHI_growthRate/MainTest.py:9]: Please add the license header to all of your python script https://github.com/ComputationalRadiationPhysics/picongpu/blob/d9ab3532a2b0d10e752b5b701ec14254f486d176/lib/python/picongpu/plugins/plot_mpl/base_visualizer.py#L1-L7

Do not forget to set the year to `2022` only and add yourself as the author.

File: lib/python/test/KHI_growthRate/MainTest.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+from time import *
+
+lt = localtime()
+date = strftime("date: %d.%m.%Y", lt)
+timeOfDay = strftime("time: %H:%M:%S", lt)
+
+try:
+	sys.path.append(sys.argv[1])
```
Comment (psychocoderHPC) [inline lib/python/test/KHI_growthRate/MainTest.py]: Why is this required? Does python has no way to get the file location of the executed file?

File: lib/python/test/KHI_growthRate/MainTest.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	from TestKHI import TestKHI
+	from ViewerKHI import ViewerKHI
+
+	field = ["total", "B_x", "B_y", "B_z", "E_x", "E_y", "E_z"]
+
+	paramDir = sys.argv[2]
```
Comment (psychocoderHPC) [inline lib/python/test/KHI_growthRate/MainTest.py]: Where should param dir points too? From the `tpl` it looks like it is the folder of `MainTest.py` which is equal to the dirname of the first parameter.

File: lib/python/test/KHI_growthRate/MainTest.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	simDir = sys.argv[3]
```
Comment (psychocoderHPC) [inline lib/python/test/KHI_growthRate/MainTest.py]: Which files must be located in `simDir`?
If you need the openPMD output maybe the prefix path + filename would be better. e.g. `simOutpur/opePMD/simData`.
The openPMD file names and folder can be chosen by the user, therefore we should allow to specify it.
IMO it would be useful to use the python argument parses instead of using unnamed fixed position parameters for the testing tool, see https://github.com/ComputationalRadiationPhysics/picongpu/blob/d9ab3532a2b0d10e752b5b701ec14254f486d176/src/tools/bin/plot_chargeConservation.py#L184

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+if [ -f !TBG_dstPath/input/bin/cuda_memtest ] && [ !TBG_numHostedGPUPerNode -eq !TBG_gpusPerNode ] ; then
+  # Run CUDA memtest to check GPU's health
+  mpiexec -np !TBG_tasks !TBG_dstPath/input/bin/cuda_memtest.sh
+else
+  echo "Note: GPU memory test was skipped as no binary 'cuda_memtest' available or compute node is not exclusively allocated. This does not affect PIConGPU, starting it now" >&2
+fi
+
+if [ $? -eq 0 ] ; then
+  # Run PIConGPU
+  mpiexec !TBG_dstPath/input/bin/picongpu !TBG_author !TBG_programParams
```
Comment (sbastrakov) [inline test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl]: Which `.cfg` files should be used to run it?

Comment (sbastrakov) [reply test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl]: I think some reasonable `.cfg` files should be included in the setup, like we do for existing examples.

File: lib/python/test/KHI_growthRate/ViewerKHI.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
+"""
```
Comment (sbastrakov) [inline lib/python/test/KHI_growthRate/ViewerKHI.py:1]: Please add `requirements.txt` for the added Python part.

File: lib/python/test/KHI_growthRate/MainTest.py (hunk 5 above)
Comment (MSVoss) [reply lib/python/test/KHI_growthRate/MainTest.py]: not really, paramDir points to the folder containing the .param files. sys.argv[0] is MainTest.py directly, [1] is the folder in which MainTest.py is located (I'm still checking your previous comment), [2] is as described above and [3] points to simOutput

File: test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl (hunk 7 above)
Comment (MSVoss) [reply test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl]: I use 16.cfg and only switch TBG_steps to something smaller like 3000

**Comment by PrometheusPi** [comment]
I had a brief discussion with @psychocoderHPC:
Please transfer the code from the `*.tpl` that runs the test to a test script `validate.sh` that returns `0` in case of passing and `1` in case of not passing the growth rate test. 
A second program called `ci.sh` should first run PIConGPU directly from the command line `picongpu -d 1 1 1 -g ... -s ... ...` as in your `submit.start` files and the call `validate.sh`. 
Both scripts should be included in a `bin` directory inside your test directory.
The reason for the separate `ci.sh` file is, that our GPU based test system does not have (yet) a scheduler and thus will run PIConGPU and later the test as in a regular script. 

We can discuss details in person after your lecture if you have time.

File: test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# PIConGPU is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with PIConGPU.
```
Comment (psychocoderHPC) [inline test/KHI_growthRate/etc/picongpu/hemera-hzdr/k80_testKHI.tpl]: Please remove the tpl file from our PR. The file will be outdated soon. You can add a `README.md` into the example where you describe what the user must do to validate the results.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

**Comment by psychocoderHPC** [comment]
@MSVoss Could you please squash your PRs into two single commits? One commit with the python stuff and one with the next ample.

sry for the late review I was always looking if the tpl files are already removed and thought the PR is not updated because the file was still in.

File: test/KHI_growthRate/bin/ci.sh [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        currentPath=$simDir/bin/
+    else
+        mkdir -p $simDir
+fi
+
+cd $currentPath/../
+pic-build -b cuda:70
+
+mkdir $simDir/simOutput
+if [ ! -d "$simDir/include/picongpu/param/" ] ; then
```
Comment (psychocoderHPC) [inline test/KHI_growthRate/bin/ci.sh]: Why is this required?
I do not understand why we need to copy the include files after PIConGPU is compiled in line 74.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

**Comment by psychocoderHPC** [comment]
@MSVoss Sry for the late response, there is only one open question before we can merge this code.

File: test/KHI_growthRate/bin/ci.sh (hunk 10 above)
Comment (MSVoss) [reply test/KHI_growthRate/bin/ci.sh]: Because I didn't want to pass another parameter to validate.sh. The copying can be removed if validate.sh gets another optional parameter giving it direction to .param files. With default simDir if not set. If this solution is better I'll change it.

File: test/KHI_growthRate/bin/ci.sh [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        echo "Destination path already in use, cannot create new folder" >&2
+        exit 1
+    fi
+fi
+
+if [ -z "$testPar" ] ; then
+        testPar=$HOME
+fi
+
+source etc/picongpu/bash-devServer-hzdr/gpu_picongpu.profile.example
```
Comment (steindev) [inline test/KHI_growthRate/bin/ci.sh]: I am not aware that the directory `bash-devServer-hzdr` exists

File: lib/python/test/KHI_growthRate/TestKHIManager.py [hunk 12]
```diff
@@ -0,0 +1,440 @@
+"""
+This file is Part of the PIConGPU+
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/TestKHIManager.py]: 2 spelling errors. Actually three. You say 'part of the PIConGPU' in every file, but it is supposed to be 'part of PIConGPU' without 'the'.

File: lib/python/test/KHI_growthRate/TestKHIManager.py [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+    getTime():
+        returns time in [w_pe ** -1]
+
+    getFieldsSumByName(key):
+        returns field data for the passed fields
+
+    getGrowthRateByName(key):
+        returns growth rate for the passed fields
+
+    getAcceptanceLevel() -> float:
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/TestKHIManager.py:43]: What is the meaning of `-> float` and why is it only written for some of the functions

File: lib/python/test/KHI_growthRate/TestKHIManager.py [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+
+    getKHIData():
+        returns the object that contains the data of the KHI
+
+    getMaxGrowthRate_Sim(field) -> float:
+        return the max growth rate of the
+        transferred field from the simulation
+    """
+
+    __acceptanceLevel = 0.2
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/TestKHIManager.py:68]: There needs to be a discussion somewhere, how this value comes about. (probably upload your thesis to zenodo and cite it here AND write a paragraph in the documentation?)

File: lib/python/test/KHI_growthRate/TestKHIManager.py [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+        """
+
+        if field is None:
+            field = self.__khiData.field
+
+        if self.__khiData.regime == "ESKHI":
+            maxValue = theoGR.getTheoryESKHI_max(
+                                            self.__khiData.gamma)
+        else:
+            maxValue = theoGR.getTheoryMIGR_max(
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/TestKHIManager.py]: Please add a comment which regime the else branch concerns

Comment (steindev) [reply lib/python/test/KHI_growthRate/TestKHIManager.py]: There is code duplication between this function and the next. Can't you call the next function in this function and then just to devision to obtain the percantage value?

File: lib/python/test/KHI_growthRate/ViewerKHI.py [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+            plt.plot(gamma, growthRate, ".k", label="value from simulation")
+
+        if plotLog:
+            plt.loglog()
+
+        plt.xlabel(r"$\gamma$")
+        plt.ylabel(r"$\Gamma[\omega_{pe}]$")
+
+        plt.legend()
+        plt.show()
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/ViewerKHI.py]: Is this script supposed to run interactively? I am wondering because the plots are not saved to disk but just shown.

File: test/KHI_growthRate/bin/ci.sh [hunk 17]
```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
```
Comment (steindev) [inline test/KHI_growthRate/bin/ci.sh:1]: First, please add a license header here as well.

Second, as this is a, maybe THE,  central file, add a little description of its purpose.

File: test/KHI_growthRate/bin/validate.sh [hunk 18]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (steindev) [inline test/KHI_growthRate/bin/validate.sh:1]: same as `ci.sh`

File: test/KHI_growthRate/bin/ci.sh (hunk 17 above)
Comment (steindev) [reply test/KHI_growthRate/bin/ci.sh:1]: Or put this description as the first lines of `--help` (preferred)

File: lib/python/test/KHI_growthRate/ViewerKHI.py [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+                    plt.plot(time, field_sum_ene[number],
+                             label="$" + lab + "$")
+
+        plt.legend()
+        plt.xlabel(r"$t[\omega_{pe}^ {-1}]$")
+        plt.ylabel(r"$\varepsilon_\mathrm{Fi}$")
+        plt.semilogy()
+        plt.grid()
+
+        plt.show()
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/ViewerKHI.py]: same as above.

File: test/KHI_growthRate/include/picongpu/param/density.param [hunk 20]
```diff
@@ -0,0 +1,46 @@
+/* Copyright 2013-2021 Axel Huebl, Heiko Burau, Rene Widera, Felix Schmitt,
```
Comment (steindev) [inline test/KHI_growthRate/include/picongpu/param/density.param]: This is an old license header (year 2021). You can also add your name, as you prepared the setup.

Comment (steindev) [reply test/KHI_growthRate/include/picongpu/param/density.param]: Same for all other files.

**Review by steindev — CHANGES_REQUESTED** [review]
Excellent that this is going to mainline!

I have a general question:
The whole test suite setup is very KHI centric at the moment. I had expected that there would be general applicable parts and KHI specific parts,where the general part can be extended with more test cases and KHI is only the first of several test cases.
Is this separation into general part and setup specific part something you want to develop with the next test cases in order to study what this general part needs to look like?
Or do you think that each test case needs to provide its own infrastructure for e.g. start, success evaluation, and garbage collection?

Second major point:
I miss documentation in our readthedocs. It does not need to be a lot. Its main purpose is to tell users that we have a test suite.
Plus a top level view of the design (probably some graphic from your Bachelor thesis), a documentation of main design decisions (like the acceptance value), and the necessary references pointing to publications with more (or all) details.
I guess this will basically be an `executive summary` of your thesis, maybe one A4 page in total.

Disclaimer: You structured the source code very well and separated implementation details in distinct files. In the review I did not proceed until the lowest level of your data structures but dived only as deep as `TestKHIManager.py`.

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 13 above)
Comment (MSVoss) [reply lib/python/test/KHI_growthRate/TestKHIManager.py:43]: As in other programming languages, the type of a return value can be specified in Python (here with -> float). However, this only makes sense where a fixed type exists and the function does not "swallow" what it gets.

File: test/KHI_growthRate/bin/ci.sh (hunk 11 above)
Comment (MSVoss) [reply test/KHI_growthRate/bin/ci.sh]: This folder exists in the dev branch.

Comment (MSVoss) [reply test/KHI_growthRate/bin/ci.sh]: But there was still an error hidden in the line. (Correction comes with all others)

File: lib/python/test/KHI_growthRate/ViewerKHI.py (hunk 16 above)
Comment (MSVoss) [reply lib/python/test/KHI_growthRate/ViewerKHI.py]: I had thought that these functions could possibly help one or the other physicist with the evaluation. Since these are only additional, I left them interactive for now. But if the possibility of storage should be given, I'll work this in. Just let me know.

**Comment by MSVoss** [comment]
I don't think that every single test case needs its own structure. However, with each new test case, I work out a generalization of the test suite. In the new test case, which I'm working on now, some classes are renamed and generalized.

**Comment by psychocoderHPC** [comment]
The CI is unhappy :-( there are some style issues in the py files.

File: test/KHI_growthRate/bin/ci.sh (hunk 17 above)
Comment (psychocoderHPC) [reply test/KHI_growthRate/bin/ci.sh:1]: note: `#!/bin/bash` must be the first line and the license header must be after it.

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 15 above)
Comment (MSVoss) [reply lib/python/test/KHI_growthRate/TestKHIManager.py]: I removed the function (it's still included in the source code in case I forgot to use the generalized one somewhere)

**Comment by MSVoss** [comment]
I hope I've corrected everything except for the documentation

**Review by psychocoderHPC — APPROVED** [review]

File: lib/python/test/KHI_growthRate/CalculateParameter.py [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+
+    growthRate(f, time, interval = None):
+        this function calculate the growthrate of a given quantity f(t)
+    """
+
+    # fixe parameter
+    __c = 299792458
+    __m_e = 9.1093837015e-31
+    __e = -1.602176634e-19
+    __e_0 = 8.8541878128e-12
```
Comment (PrometheusPi) [inline lib/python/test/KHI_growthRate/CalculateParameter.py]: Could be better taken from `scipy.constants` but fine for now

File: test/KHI_growthRate/bin/ci.sh [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+        echo "Destination path already in use, cannot create new folder" >&2
+        exit 1
+    fi
+fi
+
+if [ -z "$testPar" ] ; then
+        testPar=$HOME
+fi
+
+source $testPar/picongpu/etc/picongpu/bash-devServer-hzdr/gpu_picongpu.profile.example
```
Comment (psychocoderHPC) [inline test/KHI_growthRate/bin/ci.sh]: please remove this line. Whoever is executing this script must set up their environment before.
The CI is not running on the local dev server.

Comment (psychocoderHPC) [reply test/KHI_growthRate/bin/ci.sh]: You can assume that `pic-create` is available.

File: test/KHI_growthRate/bin/ci.sh [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+if [ "$create" == "true" ] ; then
+        mktemp -d $date_stamp
+        pic-create $testPar/picongpu/test/KHI_growthRate/ $simDir
+        currentPath=$simDir/bin/
+    else
+        mkdir -p $simDir
+fi
+
+cd $currentPath/../
+pic-build -b cuda:70
```
Comment (psychocoderHPC) [inline test/KHI_growthRate/bin/ci.sh]: please call `pic-build` only. Which backend we use and which architecture is setup by the user or the ci before

Comment (psychocoderHPC) [reply test/KHI_growthRate/bin/ci.sh]: We will execute this on CPU maybe too.

File: test/KHI_growthRate/bin/ci.sh [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+mkdir $simDir/simOutput
+
+cd $simDir/simOutput
+mpiexec -n 1 $currentPath/picongpu -d 1 1 1 -g 192 512 12 --periodic 1 1 1 -s 3000 --fields_energy.period 10
+cd ..
+
+# test with python
+bash $currentPath/validate.sh -t $testPar -s $simDir
```
Comment (psychocoderHPC) [inline test/KHI_growthRate/bin/ci.sh:91]: The tool should return the error code of the validation

```bash
bash $currentPath/validate.sh -t $testPar -s $simDir
ret=$?

if [ "$create" == "true" ] ; then
        rm -r $simDir
fi

exit $ret
```

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: lib/python/test/KHI_growthRate/KHIData.py [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+    setGrowthRateByKey(key: str, value):
+        set the growth rate of a specific field given by key
+
+    """
+
+    # fixed parameters
+    __c = 299792458
+    __m_e = 9.1093837015e-31
+    __e = -1.602176634e-19
+    __e_0 = 8.8541878128e-12
```
Comment (PrometheusPi) [inline lib/python/test/KHI_growthRate/KHIData.py]: as before, you could use scipy constants

File: lib/python/test/KHI_growthRate/ParamReader.py [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+
+                    # calculate the parameter if multiplication is included
+                    elif "*" in line[line.find("=") + 1:line.find(";")]:
+                        s = line[line.find("=") + 1:line.find(";")].split("*")
+                        for entry in s:
+                            if parameter is None:
+                                parameter = float(entry)
+                            else:
+                                if entry in '  PARAM_FACTOR':
+                                    entry = 0.15789473684210525
```
Comment (PrometheusPi) [inline lib/python/test/KHI_growthRate/ParamReader.py]: Where does this magic number come from?

Comment (PrometheusPi) [reply lib/python/test/KHI_growthRate/ParamReader.py]: Please add a comment what this is, if you want to keep this number.

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 14 above)
Comment (PrometheusPi) [reply lib/python/test/KHI_growthRate/TestKHIManager.py:68]: I agree - for now I would go for a comment.

File: lib/python/test/KHI_growthRate/TheoryGrowthRate.py [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+        gamma:  float
+                lorentz factor
+
+        return:
+        -------
+        Lambda:  float
+                 maximum growth rate of the MI
+        """
+
+        c = 299792458
```
Comment (PrometheusPi) [inline lib/python/test/KHI_growthRate/TheoryGrowthRate.py]: magic number - you could use scipy constants (or add a comment)

File: lib/python/test/KHI_growthRate/ViewerKHI.py (hunk 16 above)
Comment (PrometheusPi) [reply lib/python/test/KHI_growthRate/ViewerKHI.py]: I think storing them would be great (but as long as the storing option for the last method exist, I am fine)

File: test/KHI_growthRate/include/picongpu/param/memory.param [hunk 28]
```diff
@@ -0,0 +1,118 @@
+/* Copyright 2013-2022 Axel Huebl, Rene Widera, Benjamin Worpitz,
```
Comment (PrometheusPi) [inline test/KHI_growthRate/include/picongpu/param/memory.param]: Is this file actually needed or can we use the default?

File: test/KHI_growthRate/include/picongpu/param/radiation.param [hunk 29]
```diff
@@ -0,0 +1,195 @@
+/* Copyright 2013-2022 Rene Widera, Richard Pausch, Mika Soren Voss
```
Comment (PrometheusPi) [inline test/KHI_growthRate/include/picongpu/param/radiation.param]: This file should be removed since the test does not require a specific radiation plugin setup.

File: test/KHI_growthRate/include/picongpu/param/radiationObserver.param [hunk 30]
```diff
@@ -0,0 +1,87 @@
+/* Copyright 2013-2022 Heiko Burau, Rene Widera, Richard Pausch,
```
Comment (PrometheusPi) [inline test/KHI_growthRate/include/picongpu/param/radiationObserver.param]: same as with `radiation.param`

File: test/KHI_growthRate/include/picongpu/param/random.param [hunk 31]
```diff
@@ -0,0 +1,60 @@
+/* Copyright 2014-2022 Axel Huebl, Rene Widera
```
Comment (PrometheusPi) [inline test/KHI_growthRate/include/picongpu/param/random.param]: Are here any changes compared to the default?

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
Just some minor remarks: 
 - ignore the scipy.constant comments if you want - but since scipy is already a requirement using it would not introduce an extra dependency
  - definitely remove all param files related to radiation
  - remove param files that do not differ from the base default `./include/picongpu/param/`

File: lib/python/test/KHI_growthRate/TestKHIManager.py [hunk 32]
```diff
@@ -0,0 +1,420 @@
+"""
+This file is Part of PIConGPU.
```
Comment (steindev) [inline lib/python/test/KHI_growthRate/TestKHIManager.py]: 'Part' is still capital.

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 14 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/TestKHIManager.py:68]: I do not really understand the comment. Is there a comma missing?

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 15 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/TestKHIManager.py]: If it is not supposed to be used anymore, please remove it completely. We do not want to have dead code lying around. It will just be a source of confusion in the future.

File: lib/python/test/KHI_growthRate/ViewerKHI.py (hunk 16 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/ViewerKHI.py]: How these can be visualized interactively is definitely something for the documentation.

**Comment by steindev** [comment]
> I hope I've corrected everything except for the documentation

Documentation is left for a separate PR?

**Review by steindev — CHANGES_REQUESTED** [review]
Almost done from my perspective. Added questions to old comments and two (or so) new ones.

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 15 above)
Comment (MSVoss) [reply lib/python/test/KHI_growthRate/TestKHIManager.py]: Done

**Comment by MSVoss** [comment]
> > I hope I've corrected everything except for the documentation
> 
> Documentation is left for a separate PR?

Yes

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 14 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/TestKHIManager.py:68]: @MSVoss Don't forget to update this one as discussed last Thursday.

File: lib/python/test/KHI_growthRate/ViewerKHI.py (hunk 16 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/ViewerKHI.py]: @MSVoss Plz add this to your notes of what needs to be in the documentation. Then you can mark this as resolved.

File: lib/python/test/KHI_growthRate/ViewerKHI.py (hunk 19 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/ViewerKHI.py]: @MSVoss same as line 129 above. (By now lines 138 and 179)

File: lib/python/test/KHI_growthRate/ParamReader.py (hunk 26 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/ParamReader.py]: @MSVoss @PrometheusPi Resolved?

**Comment by steindev** [comment]
@MSVoss I checked all comments and marked those as resolved that I could evaluate. There are only a few missing that you need to check back. I believe we discussed last week something that has not been mentioned here. But I cannot remember what it was. If you do not remember as well or do not know what I mean than we will just merge.

File: lib/python/test/KHI_growthRate/TestKHIManager.py (hunk 14 above)
Comment (steindev) [reply lib/python/test/KHI_growthRate/TestKHIManager.py:68]: :+1:

**Comment by psychocoderHPC** [comment]
I will do the last review this week.

**Comment by psychocoderHPC** [comment]
@MSVoss ~~could you please check the following formating/import issues~~

The formatting issue is in the third-party folder, please rebase your PR against the latest dev to solve the problems.

**Review by psychocoderHPC — APPROVED** [review]

**Comment by psychocoderHPC** [comment]
@MSVoss thx for the updates, from my side all, looks fine!

@steindev @PrometheusPi could you please recheck this PR again until tomorrow so that we can merge it soon?

**Comment by PrometheusPi** [comment]
@psychocoderHPC will do

**Review by PrometheusPi — APPROVED** [review]

**Review by steindev — APPROVED** [review]
